codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java

```diff
@@ -0,0 +1,310 @@
+        allocHandle.reset(config());
+        if (inboundBuffer.isEmpty()) {
+            readInProgress = true;
+            return;
+        }
+
+        do {
+            Object m = inboundBuffer.poll();
+            if (m == null) {
+                break;
+            }
+            if (!doRead0(m, allocHandle)) {
+                // Channel closed, and already cleaned up.
+                return;
+            }
+        } while (allocHandle.continueReading());
+
+        allocHandle.readComplete();
+        pipeline().fireChannelReadComplete();
+    }
+
+    @Override
+    protected final void doWrite(ChannelOutboundBuffer in) throws Exception {
+        if (closed) {
+            throw CLOSED_CHANNEL_EXCEPTION;
+        }
+
+        EventExecutor preferredExecutor = preferredEventExecutor();
+
+        if (preferredExecutor.inEventLoop()) {
+            for (;;) {
+                Object msg = in.current();
+                if (msg == null) {
+                    break;
+                }
+                try {
+                    doWrite(ReferenceCountUtil.retain(msg));
+                } catch (Throwable t) {
+                    // It would be nice to fail the future, but we can't do that if not on the event
+                    // loop. So we instead opt for a solution that is consistent.
+                    pipeline().fireExceptionCaught(t);
+                }
+                in.remove();
+            }
+            doWriteComplete();
+        } else {
+            // Use a copy because the original msgs will be recycled by AbstractChannel.
+            final Object[] msgsCopy = new Object[in.size()];
+            for (int i = 0; i < msgsCopy.length; i ++) {
+                msgsCopy[i] = ReferenceCountUtil.retain(in.current());
+                in.remove();
+            }
+
+            preferredExecutor.execute(new OneTimeTask() {
+                @Override
+                public void run() {
+                    for (Object msg : msgsCopy) {
+                        try {
+                            doWrite(msg);
+                        } catch (Throwable t) {
+                            pipeline().fireExceptionCaught(t);
+                        }
+                    }
+                    doWriteComplete();
+                }
+            });
+        }
+    }
+
+    /**
+     * Process a single write. Guaranteed to eventually be followed by a {@link #doWriteComplete()},
+     * which denotes the end of the batch of writes. May be called from any thread.
+     */
+    protected abstract void doWrite(Object msg) throws Exception;
+
+    /**
+     * Process end of batch of {@link #doWrite()}s. May be called from any thread.
+     */
+    protected abstract void doWriteComplete();
+
+    /**
+     * The ideal thread for events like {@link #doWrite()} to be processed on. May be used for
+     * efficient batching, but not required.
+     */
+    protected abstract EventExecutor preferredEventExecutor();
```

Comment:
It's sort of any communication with the parent channel. Today, that _could_ include `bytesConsumed()`. The only reason I'm not using it for `bytesConsumed()` is because I can't do any batching.

However, with the next iteration it'll probably be communicating exclusively through messages at which point only `write` and `flush` would be called. I could rename it now, but I think I'll wait and see how things turn out.